### PR TITLE
UCP/MM: Selectively exclude MDs from rcache

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/ucp/core/ucp_mm.c [hunk 1]
```diff
@@ -1014,6 +1022,13 @@ static ucs_rcache_ops_t ucp_mem_rcache_ops = {
 ucs_status_t ucp_mem_rcache_init(ucp_context_h context)
 {
     ucs_rcache_params_t rcache_params;
+    ucs_status_t status;
+    ucp_md_map_t dummy;
+
+    status = ucp_mm_get_alloc_md_map(context, &dummy);
+    if (status != UCS_OK) {
+        return status;
+    }
```
Comment (brminich) [inline src/ucp/core/ucp_mm.c:1031]: why is this needed?

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -712,7 +712,9 @@ enum {
     UCT_MD_FLAG_SOCKADDR   = UCS_BIT(7),  /**< MD support for client-server
                                                connection establishment via
                                                sockaddr */
-    UCT_MD_FLAG_INVALIDATE = UCS_BIT(8)   /**< MD supports memory invalidation */
+    UCT_MD_FLAG_INVALIDATE = UCS_BIT(8),  /**< MD supports memory invalidation */
+    UCT_MD_FLAG_DONT_CACHE = UCS_BIT(9)   /**< MD registrations should not be
+                                               cached between operation */
```
Comment (brminich) [inline src/uct/api/uct.h:717]: `operations`?

File: src/ucp/core/ucp_context.h [hunk 3]
```diff
@@ -263,6 +263,10 @@ typedef struct ucp_context {
        ucp_mem_map() will register memory for all those domains. */
     ucp_md_map_t                  reg_md_map[UCS_MEMORY_TYPE_LAST];
 
+    /* Map of MDs that for host memory should be excluded from ucp_mem_map()
+       and not be cached between operation. */
+    ucp_md_map_t                  dont_cache_md_map;
```
Comment (brminich) [inline src/ucp/core/ucp_context.h:268]: maybe `exclude_md_map`?

File: src/ucp/core/ucp_mm.inl [hunk 4]
```diff
@@ -78,6 +78,14 @@ ucp_memh_put(ucp_context_h context, ucp_mem_h memh, int invalidate)
                 (ucs_rcache_invalidate_comp_func_t)ucs_empty_function, NULL);
     }
 
+    if ((memh->super.flags & UCS_RCACHE_REGION_FLAG_PGTABLE) &&
+        (memh->super.refcount == 2) &&
```
Comment (brminich) [inline src/ucp/core/ucp_mm.inl:82]: i'd define some constant instead of `2`, or plz add some comment explaining the value

File: src/ucp/core/ucp_mm.c (hunk 1 above)
Comment (Artemy-Mellanox) [reply src/ucp/core/ucp_mm.c:1031]: There is an issue with cuMemAlloc - we need to call cuda register once to initialize cuda context.

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -712,7 +712,9 @@ enum {
     UCT_MD_FLAG_SOCKADDR   = UCS_BIT(7),  /**< MD support for client-server
                                                connection establishment via
                                                sockaddr */
-    UCT_MD_FLAG_INVALIDATE = UCS_BIT(8)   /**< MD supports memory invalidation */
+    UCT_MD_FLAG_INVALIDATE = UCS_BIT(8),  /**< MD supports memory invalidation */
+    UCT_MD_FLAG_WEAK_CACHE = UCS_BIT(9)   /**< MD registrations should not be
```
Comment (yosefe) [inline src/uct/api/uct.h:716]: send options to mailing list:
Option1: "negative" flag: UCT_MD_FLAG_WEAK_CACHE / UCT_MD_FLAG_DONT_CACHE
Option2: "positive" flag: UCT_MD_FLAG_LEAVE_PINNED / UCT_MD_FLAG_ALLOW_CACHE /...

File: src/ucp/core/ucp_mm.c [hunk 6]
```diff
@@ -429,24 +442,88 @@ ucp_memh_get_slow(ucp_context_h context, void *address, size_t length,
...
+
+    memh->mem_type = mem_type;
+    status         = ucp_memh_register(context, memh,
+                                       reg_md_map & context->weak_md_map,
+                                       address, length, uct_flags);
+    if (status != UCS_OK) {
+        goto err_free_memh;
+    }
+
+    status = ucp_memh_get_slow(context, address, length, mem_type,
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:496]: get w/o slow

File: src/ucp/core/ucp_mm.c [hunk 7]
```diff
@@ -429,24 +442,88 @@ ucp_memh_get_slow(ucp_context_h context, void *address, size_t length,
...
+        ucs_free(memh);
+    } else {
+        ucs_rcache_region_put_unsafe(context->rcache, &memh->super);
+    }
+
+    goto out;
+}
+
+ucs_status_t
+ucp_memh_get_weak(ucp_context_h context, void *address, size_t length,
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:469]: pls add comments to this func

File: src/ucp/core/ucp_mm.c [hunk 8]
```diff
@@ -429,24 +442,88 @@ ucp_memh_get_slow(ucp_context_h context, void *address, size_t length,
...
+    }
+
+    status = ucp_memh_get_slow(context, address, length, mem_type,
+                               reg_md_map & ~memh->md_map,
+                               uct_flags, &memh->parent);
+    if (status != UCS_OK) {
+        goto err_dereg;
+    }
+
+    ucs_for_each_bit(md_index, memh->parent->md_map & ~memh->md_map) {
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:503]: copy just reg_md_map

File: src/ucp/core/ucp_mm.c [hunk 9]
```diff
@@ -783,7 +865,7 @@ ucp_mpool_free(ucp_worker_h worker, ucs_mpool_t *mp, void *chunk)
     ucp_mem_desc_t *chunk_hdr;
 
     chunk_hdr = (ucp_mem_desc_t*)chunk - 1;
-    ucp_memh_put(worker->context, chunk_hdr->memh, 0);
+    ucp_memh_put(worker->context, chunk_hdr->memh, 1);
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:868]: ucp_memh_alloc should also create "indirect" memh , therefore would not need to pass invalidate=1 here, because such "indirect" memh would be always released fully by ucp_memh_put
probably could also remove map_refcount by this method

File: src/ucp/core/ucp_mm.c [hunk 10]
```diff
@@ -763,6 +844,7 @@ ucp_mpool_malloc(ucp_worker_h worker, ucs_mpool_t *mp, size_t *size_p, void **ch
 
     status = ucp_memh_alloc(worker->context, NULL,
                             *size_p + sizeof(*chunk_hdr), UCS_MEMORY_TYPE_HOST,
+                            worker->context->reg_md_map[UCS_MEMORY_TYPE_HOST],
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:847]: can we remove this param?

File: src/ucp/core/ucp_mm.h [hunk 11]
```diff
@@ -32,6 +32,9 @@ typedef struct ucp_mem {
     ucp_md_index_t      alloc_md_index; /* Index of MD used to allocated the memory */
     ucp_md_map_t        md_map;         /* Which MDs have valid memory handles */
     uint32_t            map_count;      /* ucp_mem_map/unmap referrence count */
+    struct ucp_mem      *parent;        /* NULL if entry should be returned to rcache
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.h:35]: uct_mem_h
https://github.com/openucx/ucx/blob/b6d73b7a84c4380cc52b449bf95ed6ece8a706da/src/ucp/api/ucp_def.h#L167

File: src/ucp/core/ucp_mm.c [hunk 12]
```diff
@@ -377,6 +376,28 @@ static ucs_status_t ucp_memh_register(ucp_context_h context, ucp_mem_h memh,
...
+            (sizeof(uct_mem_h) * context->num_mds), "ucp_rcache");
+    if (memh == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    memh->super.super.start = (uintptr_t)address;
+    memh->super.super.end   = (uintptr_t)address + length;
+    memh->alloc_md_index    = UCP_NULL_RESOURCE;
+    memh->alloc_method      = UCT_ALLOC_METHOD_LAST;
+    memh->parent            = memh;
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:394]: add comment why

File: test/gtest/ucp/test_ucp_mmap.cc [hunk 13]
```diff
@@ -455,8 +455,12 @@ UCS_TEST_P(test_ucp_mmap, rereg)
                                              &test_memh);
                 ASSERT_UCS_OK(status);
 
-                // Check that the same memory handle was returned from RCACHE
-                EXPECT_EQ(memh, test_memh);
+                if (!mem_buffer::is_gpu_supported()) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_mmap.cc:458]: 1. inverse order
2. check memh->md_map instead of is_gpu_supported

**Comment by yosefe** [comment]
GPU failure seems relevant
```
2022-08-08T04:26:41.4414131Z [ RUN      ] ud/test_ucp_am_nbx_eager_memtype.basic/7 <ud_v,cuda_copy,rocm_copy/prereg/proto/cuda/host>
2022-08-08T04:26:41.6186332Z [1659932801.618056] [swx-rdmz-ucx-gpu-02:11452:0]          rcache.c:1291 UCX  WARN  rcache mlx5_0: 1 regions remained on lru list, first region: 0x50ddbb0
2022-08-08T04:26:41.6219225Z [1659932801.621584] [swx-rdmz-ucx-gpu-02:11452:0]          rcache.c:1291 UCX  WARN  rcache mlx5_1: 1 regions remained on lru list, first region: 0x50ddd10
2022-08-08T04:26:41.6245059Z /__w/1/s/contrib/../test/gtest/common/test.cc:366: Failure
2022-08-08T04:26:41.6246084Z Failed
2022-08-08T04:26:41.6246944Z Got 0 errors and 2 warnings during the test
2022-08-08T04:26:41.6248172Z [     INFO ] < /__w/1/s/contrib/../src/ucs/memory/rcache.c:1291 rcache mlx5_0: 1 regions remained on lru list, first region: 0x50ddbb0 >
2022-08-08T04:26:41.6249554Z [     INFO ] < /__w/1/s/contrib/../src/ucs/memory/rcache.c:1291 rcache mlx5_1: 1 regions remained on lru list, first region: 0x50ddd10 >
2022-08-08T04:26:41.6250912Z [  FAILED  ] ud/test_ucp_am_nbx_eager_memtype.basic/7, where GetParam() = ud_v,cuda_copy,rocm_copy/prereg/proto/cuda/host (184 ms)
```

File: src/ucp/core/ucp_mm.c [hunk 14]
```diff
@@ -475,34 +474,48 @@ ucp_memh_init_from_mem(ucp_context_h context, ucp_mem_h memh,
...
-                              reg_md_map, uct_flags, &memh->parent);
+        status = ucp_memh_get(context, address, length, mem_type, cache_md_map,
+                              uct_flags, &memh->parent);
         if (status != UCS_OK) {
-            return status;
+            goto err;
         }
 
-        ucs_for_each_bit(md_index, reg_md_map) {
+        ucs_for_each_bit(md_index, memh->parent->md_map) {
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:500]: shall we expect cache_md_map == memh->parent->md_map? if yes pls assertv it

File: src/ucp/core/ucp_mm.c [hunk 15]
```diff
@@ -325,9 +325,8 @@ void ucp_memh_cleanup(ucp_context_h context, ucp_mem_h memh)
 
     /* Have a parent memory handle from rcache */
     if ((memh->parent != NULL) && (memh->parent != memh)) {
-        /* TODO remove invalidate, replace with ucp_memh_dereg weak MDs */
-        ucs_rcache_region_invalidate(context->rcache, &memh->parent->super,
-                (ucs_rcache_invalidate_comp_func_t)ucs_empty_function, NULL);
+        md_map &= ~memh->parent->md_map;
+        ucp_memh_dereg(context, memh, md_map);
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:329]: better pass `md_map & ~memh->parent->md_map` instead of updating md_map

File: src/uct/cuda/cuda_copy/cuda_copy_md.c [hunk 16]
```diff
@@ -49,6 +49,8 @@ static ucs_status_t uct_cuda_copy_md_query(uct_md_h md, uct_md_attr_t *md_attr)
     md_attr->cap.reg_mem_types    = UCS_BIT(UCS_MEMORY_TYPE_HOST) |
                                     UCS_BIT(UCS_MEMORY_TYPE_CUDA) |
                                     UCS_BIT(UCS_MEMORY_TYPE_CUDA_MANAGED);
+    md_attr->cap.cache_mem_types  = UCS_BIT(UCS_MEMORY_TYPE_CUDA) |
```
Comment (yosefe) [inline src/uct/cuda/cuda_copy/cuda_copy_md.c:52]: shouldn't this be 0?

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.c [hunk 17]
```diff
@@ -33,6 +33,7 @@ static ucs_status_t uct_cuda_ipc_md_query(uct_md_h md, uct_md_attr_t *md_attr)
                                     UCT_MD_FLAG_NEED_RKEY |
                                     UCT_MD_FLAG_INVALIDATE;
     md_attr->cap.reg_mem_types    = UCS_BIT(UCS_MEMORY_TYPE_CUDA);
+    md_attr->cap.cache_mem_types  = UCS_BIT(UCS_MEMORY_TYPE_CUDA);
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_md.c:36]: shouldn't this be 0?

File: src/uct/api/uct.h [hunk 18]
```diff
@@ -1436,6 +1436,7 @@ struct uct_md_attr {
         size_t               max_reg;   /**< Maximal registration size */
         uint64_t             flags;     /**< UCT_MD_FLAG_xx */
         uint64_t             reg_mem_types; /**< Bitmap of memory types that Memory Domain can be registered with */
+        uint64_t             cache_mem_types; /**< Bitmap of memory types that should be cached for this Memory Domain */
```
Comment (yosefe) [inline src/uct/api/uct.h:1439]: 1. should -> could
2. add it to ucx_info

File: src/ucp/dt/datatype_iter.inl [hunk 19]
```diff
@@ -507,6 +507,9 @@ ucp_datatype_iter_mem_reg(ucp_context_h context, ucp_datatype_iter_t *dt_iter,
 {
     if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_CONTIG, dt_mask)) {
         ucs_assert(dt_iter->type.contig.memh == NULL);
+        /* Iterator may work with cacheable MDs only */
+        ucs_assert(ucs_test_all_flags(
```
Comment (yosefe) [inline src/ucp/dt/datatype_iter.inl:511]: assertv

File: src/ucp/dt/datatype_iter.inl [hunk 20]
```diff
@@ -508,8 +508,12 @@ ucp_datatype_iter_mem_reg(ucp_context_h context, ucp_datatype_iter_t *dt_iter,
     if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_CONTIG, dt_mask)) {
         ucs_assert(dt_iter->type.contig.memh == NULL);
         /* Iterator may work with cacheable MDs only */
-        ucs_assert(ucs_test_all_flags(
-                    context->cache_md_map[dt_iter->mem_info.type], md_map));
+        ucs_assertv(ucs_test_all_flags(
+                    context->cache_md_map[dt_iter->mem_info.type], md_map),
+                    "iterator mem_type %d got md_map 0x%" PRIx64
```
Comment (yosefe) [inline src/ucp/dt/datatype_iter.inl:513]: `mem_type=%s cache_md_map=%PRIx64 md_map=%PRIx64, ucs_memory_type_names[..], ...`

File: test/gtest/ucp/test_ucp_proto.cc [hunk 21]
```diff
@@ -57,6 +57,7 @@ ucp_md_map_t test_ucp_proto::get_md_map(ucs_memory_type_t mem_type)
         const uct_md_attr_t *md_attr = &context()->tl_mds[md_index].attr;
         if ((md_attr->cap.flags & UCT_MD_FLAG_REG) &&
             (md_attr->cap.reg_mem_types & UCS_BIT(mem_type)) &&
+            (md_attr->cap.cache_mem_types & UCS_BIT(mem_type)) &&
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_proto.cc:60]: why needed?

File: src/uct/cuda/cuda_copy/cuda_copy_md.c (hunk 16 above)
Comment (yosefe) [reply src/uct/cuda/cuda_copy/cuda_copy_md.c:52]: unresolved

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.c (hunk 17 above)
Comment (yosefe) [reply src/uct/cuda/cuda_ipc/cuda_ipc_md.c:36]: unresolved

File: src/uct/api/uct.h (hunk 18 above)
Comment (yosefe) [reply src/uct/api/uct.h:1439]: unresolved

File: test/gtest/ucp/test_ucp_proto.cc (hunk 21 above)
Comment (Artemy-Mellanox) [reply test/gtest/ucp/test_ucp_proto.cc:60]: it is used to test `ucp_datatype_iter_mem_reg()` which only work for cacheable MDs

Comment (yosefe) [reply test/gtest/ucp/test_ucp_proto.cc:60]: why is it so?
will it also be the case for proto v2 and when we replace dt_reg by ucp_memh for protov1?

Comment (Artemy-Mellanox) [reply test/gtest/ucp/test_ucp_proto.cc:60]: yes, that is design
`ucp_datatype_iter_mem_reg()` call `ucp_memh_get()` which get memh from rcache and we don't want _selectively exceluded MDs_ in rcache.

Comment (yosefe) [reply test/gtest/ucp/test_ucp_proto.cc:60]: `/* ucp_datatype_iter_mem_reg() always goes directly to registration cache */`

File: src/tools/info/tl_info.c [hunk 22]
```diff
@@ -460,6 +462,35 @@ static void print_md_info(uct_component_h component,
...
             printf("#           memory invalidation is supported\n");
         }
+
+        buf[0] = '\0';
+        ucs_memory_type_for_each(mem_type) {
+            if (!(UCS_BIT(mem_type) & (md_attr.cap.reg_mem_types |
+                                       md_attr.cap.alloc_mem_types))) {
+                continue;
+            }
+            strncat(buf, ucs_memory_type_names[mem_type],
```
Comment (yosefe) [inline src/tools/info/tl_info.c:472]: space line before

File: src/tools/info/tl_info.c [hunk 23]
```diff
@@ -387,8 +387,10 @@ static void print_md_info(uct_component_h component,
                           ucs_config_print_flags_t print_flags,
                           const char *req_tl_name)
 {
+    char buf[256] = {0};
```
Comment (yosefe) [inline src/tools/info/tl_info.c:390]: maybe use string_buffer?

**Review by yosefe — DISMISSED** [review]

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
